Storage/Logins.swift

```diff
@@ -0,0 +1,239 @@
+    var passwordField: String? { get set }
+
+    func toDict() -> [String: String]
+}
+
+public protocol LoginUsageData: LoginData {
+    var timeCreated: MicrosecondTimestamp { get set }
+    var timeLastUsed: MicrosecondTimestamp { get set }
+    var timePasswordChanged: MicrosecondTimestamp { get set }
+}
+
+public protocol SyncableLoginData {
+    var guid: String? { get set }
+    var isDeleted: Bool { get set }
+}
+
+public class Login : Printable, SyncableLoginData, LoginUsageData, Equatable {
+    public let credentials: NSURLCredential
+    public let protectionSpace: NSURLProtectionSpace
+
+    public var hostname: String { return protectionSpace.host }
+    public var username: String? { return credentials.user }
+    public var password: String { return credentials.password! }
+    private var _httpRealm: String? = nil
+    public var usernameField: String? = nil
+    public var passwordField: String? = nil
+
+    public var httpRealm: String? {
+        get { return self._httpRealm ?? protectionSpace.realm }
+        set { self._httpRealm = newValue }
+    }
+
+    public var formSubmitUrl: String? = nil {
+        didSet {
+            if self.formSubmitUrl == nil || self.formSubmitUrl!.isEmpty {
+                return
+            }
+
+            let url2 = NSURL(string: self.hostname)
+            let url1 = NSURL(string: self.formSubmitUrl!)
+            if url1?.host != url2?.host {
+                assertionFailure("Form submit url domain doesn't match login's domain")
+                formSubmitUrl = nil
+            }
+        }
+    }
+
+    // SyncableLoginData
+    public var guid: String? = nil
+    public var isDeleted: Bool = false
+
+    // UsageLignData
+    public var timeCreated = NSDate.nowMicroseconds()
+    public var timeLastUsed = NSDate.nowMicroseconds()
+    public var timePasswordChanged = NSDate.nowMicroseconds()
+
+    // Printable
+    public var description: String {
+        return "Login: \(username) for \(hostname)"
+    }
+
+    public class func createWith(hostname: String, username: String, password: String) -> LoginData {
+        return Login(hostname: hostname, username: username, password: password) as LoginData
+    }
+
+    public class func createWith(credential: NSURLCredential, protectionSpace: NSURLProtectionSpace) -> LoginData {
+        return Login(credential: credential, protectionSpace: protectionSpace) as LoginData
+    }
+
+    init(hostname: String, username: String, password: String) {
+        self.credentials = NSURLCredential(user: username, password: password, persistence: NSURLCredentialPersistence.None)
+        self.protectionSpace = NSURLProtectionSpace(host: hostname, port: 0, `protocol`: nil, realm: nil, authenticationMethod: nil)
+    }
+
+    init(credential: NSURLCredential, protectionSpace: NSURLProtectionSpace) {
+        self.credentials = credential
+        self.protectionSpace = protectionSpace
+    }
+
+    public func toDict() -> [String: String] {
```

Comment:
Would it be better to have this return `[String: String?]` and avoid all of the `?? ""`s?